storages/multipart_round_tripper.go

```diff
@@ -2,111 +2,104 @@ package storages
-
-	multiPartRoundTripper := &MultiPartRoundTripper{
-		fallBackRoundTripper: cluster.transport,
-		syncLog:              syncLog,
-	}
-	log.Debug("NewMultipart ", cluster.Name(), " ", cluster.name, " ", len(cluster.Backends()))
-	multiPartRoundTripper.setupRoundTripper(cluster.Backends())
-	return multiPartRoundTripper
-}
-
-func (multiPartRoundTripper *MultiPartRoundTripper) setupRoundTripper(backends []http.RoundTripper) {
+// Cancel Client interface
+func (multiPartRoundTripper MultiPartRoundTripper) Cancel() error { return nil }
 
+// newMultiPartRoundTripper initializes multipart client
+func newMultiPartRoundTripper(backends []*Backend) client {
+	multiPartRoundTripper := &MultiPartRoundTripper{}
 	var backendsEndpoints []string
 	var activeBackendsEndpoints []string
 
 	multiPartRoundTripper.backendsRoundTrippers = make(map[string]*Backend)
 
-	for _, roundTripper := range backends {
-
-		if backend, isBackendType := roundTripper.(*Backend); isBackendType {
-
-			if !backend.Maintenance {
-				multiPartRoundTripper.backendsRoundTrippers[backend.Endpoint.Host] = backend
-				activeBackendsEndpoints = append(activeBackendsEndpoints, backend.Endpoint.Host)
-			}
-
-			backendsEndpoints = append(backendsEndpoints, backend.Endpoint.Host)
+	for _, backend := range backends {
+		if !backend.Maintenance {
+			multiPartRoundTripper.backendsRoundTrippers[backend.Endpoint.Host] = backend
+			activeBackendsEndpoints = append(activeBackendsEndpoints, backend.Endpoint.Host)
 		}
+
+		backendsEndpoints = append(backendsEndpoints, backend.Endpoint.Host)
 	}
 
 	multiPartRoundTripper.backendsEndpoints = backendsEndpoints
 	multiPartRoundTripper.backendsRing = hashring.New(activeBackendsEndpoints)
+	return multiPartRoundTripper
 }
 
-// RoundTrip performs a RoundTrip using the strategy described in MultiPartRoundTripper
-func (multiPartRoundTripper *MultiPartRoundTripper) RoundTrip(request *http.Request) (response *http.Response, requestError error) {
-
-	if isMultiPartUploadRequest(request) {
-
-		if !multiPartRoundTripper.canHandleMultiUpload() {
-			log.Debugf("Multi upload for %s failed - no backends available.", request.URL.Path)
-			return nil, errors.New("Can't handle multi upload")
-		}
-
-		multiUploadBackend, backendSelectError := multiPartRoundTripper.pickBackend(request.URL.Path)
-
-		if backendSelectError != nil {
-			log.Debugf("Multi upload failed for %s - %s", backendSelectError, request.URL.Path)
-			return nil, errors.New("Can't handle multi upload")
-		}
+// Do performs backend request
+func (multiPartRoundTripper *MultiPartRoundTripper) Do(request *http.Request) <-chan BackendResponse {
+	out := make(chan BackendResponse)
+	if !multiPartRoundTripper.canHandleMultiUpload() {
+		log.Debugf("Multi upload for %s failed - no backends available.", request.URL.Path)
+		go func() {
+			out <- BackendResponse{Response: nil, Error: errors.New("Can't handle multi upload")}
```

Comment:
I'd say that this shouldn't  be treated as an error, because it's kinda confusing. The reason that we have to sync is because some other backend was picked to handle the multi upload, not because there was an error. Imho the message `Can't handle multi upload` is misleading. Can we instead use the `Response` struct, the the backend wasn't picked for multipart and thus we need to sync it?